elastic/apm-server
# [TBS] Wait for ProcessBatch before stopping sampler

## Motivation/summary

prevent data loss from stopping the tail-based sampler while batches are still being processed

## Checklist


## How to test these changes

I wasn't able to reproduce the issue as described in https://github.com/elastic/apm-server/issues/8339, ie. while sending a continuous stream of transactions and then disabling TBS, I never saw `processor is stopped` in the logs.

With the change I added, I did see the log message indicating that a batch was received in the TBS while it was stopping, `received batch while stopping; waiting`; after the configured timeout the "no batches received" message was logged and the TBS stopped.

1. start the stack
    ```
    ~/w/apm-integration-testing (main=)
    # ./scripts/compose.py start 8.4.0 --remove-orphans --no-apm-server
    ```
2. start apm-server with tbs enabled
3. send a stream of transactions to the apm-server. I used the following go program:

```go
package main

import (
	"context"
	"fmt"
	"os"

	"go.elastic.co/apm/v2"
)

func main() {
	version := "undefined"
	if len(os.Args) > 1 {
		version = os.Args[1]
	}
	ctx := context.Background()
	name := fmt.Sprintf("apm-server-%s", version)
	tracer := apm.DefaultTracer()
	for {
		tx := tracer.StartTransaction("tx1", "type")
		ctx = apm.ContextWithTransaction(ctx, tx)
		span, ctx := apm.StartSpan(ctx, "span1", "type")
		span.Context.SetLabel("enabled", true)
		span.Context.SetLabel("string", "string")
		span.Context.SetLabel("string2", "string2")
		span.Context.SetLabel("string3", "string3")
		span.Context.SetLabel("numbers", 1234.56)
		span.End()
		tx.End()

		tx = tracer.StartTransaction("tx2", "type")
		ctx = apm.ContextWithTransaction(ctx, tx)
		span, ctx = apm.StartSpan(ctx, "span2", "type")
		span.Context.SetLabel("other", "label")
		span.End()
		tx.End()

		tracer.Flush(nil)
	}
	fmt.Printf("%s: %+v\n", name, tracer.Stats())
}
```
4. while still streaming transactions, disable tbs and reload the apm-server
5. check the log output for something like this:
```
{"log.level":"info","@timestamp":"2022-06-15T14:44:31.059+0200","log.logger":"sampling","log.origin":{"file.name":"sampling/processor.go","file.line":328},"message":"received batch while stopping; waiting","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-06-15T14:44:36.059+0200","log.logger":"sampling","log.origin":{"file.name":"sampling/processor.go","file.line":335},"message":"no batches received; stopping","service.name":"apm-server","ecs.version":"1.6.0"}
```

Note: I ran this locally using https://github.com/elastic/apm-server/commit/fb74c6e3048ebf33e1a6a8e8f0fad767d6e836c2, so my actual flow was to update `apm-server.dev.yml` and `kill -s USR1 $(pidof apm-server)`, but the update via managed apm-server + apm-integration should be the same.

## Related issues

closes https://github.com/elastic/apm-server/issues/8339
